py: Allow CC=gcc-4.7 to be specified and have 32-bit builds still work

dhylands opened the pull request:
Checking this patch to see if it fixes the travis breakage.

The issue is that when CC=gcc-4.7 is specified on the make command line, the CC += -m32 is ignored.

dhylands committed:
py: Allow CC=gcc-4.7 to be specified and have 32-bit builds still works

stinos commented:
> Checking this patch to see if it fixes the travis breakage

Just pushing to any branch of your own fork should already start a travis build, no need to create a PR for it: https://travis-ci.org/dhylands/micropython/builds/

dhylands commented:
> Checking this patch to see if it fixes the travis breakage
> Just pushing to any branch of your own fork should already start a travis build, no need to create a PR for it: https://travis-ci.org/dhylands/micropython/builds/

You're right - for some reason I hadn't made the mental connection. I setup a custom branch in my own repo for doing travis builds on a bunch of additional targets whenever the repository changes and hadn't noticed my own pushes intermingled in there.

Next time I'll know better.

dhylands commented:
Yes - that seems to have fixed it.

if you want nanbox to be built with FFI, then we'll need to add libffi-dev:i386 to the list of installed packages and remove the MICROPY_PY_FFI=0 addition that I made to the unix/Makefile

dpgeorge commented:
Thanks @dhylands.  I was trying to reproduce the Travis failure locally but didn't manage to.

Do you think using override is the "right thing" to do?  Other option would be to add -m32 to CFLAGS and LDFLAGS.  But actually I think what you did is simpler.

Yes it would be good to have nanbox with modffi.

dhylands commented:
> Do you think using override is the "right thing" to do? Other option would be to add -m32 to CFLAGS and LDFLAGS. But actually I think what you did is simpler.

There are probably a couple of other ways of doing this:

1 - use override (as I coded it). This probably warrants a comment about why the use of override is needed.

2 - Assign CFLAGS and LDFLAGS in mkenv.mk and fix all of the other Makefiles to use += and not assign to CFLAGS and LDFLAGS

3 - Create a new variable, say something like M32_FLAGS, and have appropriate places (currently it looks like unix/Makefile and windows/Makefile) include M32_FLAGS into the CFLAGS/LDFLAGS appropriately.

I kind of like option 3. I went with option 1 because @pfalcon decided that -m32 should be bundled with CC, so I was just carrying that along (and it also happens to be the simplest solution).

I added some comments to explain why we're using override.

I also fixed up nanbox so that it build with ffi, and added libffi-dev:i386 to the list of packages retrieved for travis.

gcc-4.7-multilib was needed to get the i386 version of libgcc

pfalcon commented:
> Other option would be to add -m32 to CFLAGS and LDFLAGS.

I moved it from CFLAGS/LDFLAGS some time ago, because it's not really a compile mode option, it's compile**r** selection option. Specific cause for the change was support for building 3rd-party libs in lib/ - these should use the same compiler, but compile options are generally differ. Here's for comparison my question about -m32 for arm64: https://lists.96boards.org/pipermail/dev/2015-November/000544.html .

So, general solution to the above problem is that 32-bit build with gcc-4.7 should use CC="gcc-4.7 -m32" (if not CC="i386-linux-gcc-4.7", as the mail above suggests).

dhylands commented:
So then, what's the purpose of MICROPY_FORCE_32BIT. It seems confusing to have 2 potential ways of saying build 32-bit and have to use one way sometimes and one way another?

If we're going to say that you should build 32-bit by passing in -m32 as part of CC, then we should get rid of MICROPY_FORCE_32BIT (at least from mpconfigport.mk) and see if -m32 is present in CC in order to determine if we're building 32-bit (for the Makefiles wihch need to know)

pfalcon commented:
> So then, what's the purpose of MICROPY_FORCE_32BIT.

You added it, and I personally find it useful. It's easy to use "gcc -m32" - if you know about it. If you just started with 64-bit systems, it's not obvious. And as mail quote above shows, it's no going to be scalable to other 64-bit systems but x86. I'd personally leave everything as is for now. Well, if you think that adding these "override"'s is right thing, can do. Otherwise, build failure was fixed by #1786.

dpgeorge commented:
I'd say this PR is no longer needed now that things work with Travis.  I use MICROPY_FORCE_32BIT a lot but never ran into an issue where I needed to change CC as well.
